### DISCUSSION TITLE
Adding a new child node to a displayed scene causes crash

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1278] [General]
Before I create a minimum reproducible example, I thought I'd ask if I'm doing something fundamentally wrong.

What I did:

1. I created a scene graph with 4 VertexDraw nodes below the root group which I'm displaying in a `vsgQt::Viewer`..
2. Based on interaction, I added one more node to my displayed scene.

The result was a crash. Here is a stack trace from Valgrind:

```
==1770777== Invalid read of size 8
==1770777==    at 0x70AB52: vsg::ref_ptr<vsg::GraphicsPipeline::Implementation>::operator->() const (ref_ptr.h:171)
==1770777==    by 0x70A758: vsg::GraphicsPipeline::vk(unsigned int) const (GraphicsPipeline.h:63)
==1770777==    by 0x70A5D6: vsg::BindGraphicsPipeline::record(vsg::CommandBuffer&) const (GraphicsPipeline.cpp:278)
==1770777==    by 0x7B3218: vsg::StateStack<vsg::StateCommand>::record(vsg::CommandBuffer&) (State.h:68)
==1770777==    by 0x7B2671: vsg::State::record() (State.h:292)
==1770777==    by 0x7B0484: vsg::RecordTraversal::apply(vsg::VertexIndexDraw const&) (RecordTraversal.cpp:341)
==1770777==    by 0x85F356: vsg::Inherit<vsg::Command, vsg::VertexIndexDraw>::accept(vsg::RecordTraversal&) const (Inherit.h:72)
==1770777==    by 0x6EA83D: vsg::StateGroup::traverse(vsg::RecordTraversal&) const (StateGroup.h:77)
==1770777==    by 0x7B0A65: vsg::RecordTraversal::apply(vsg::StateGroup const&) (RecordTraversal.cpp:459)
==1770777==    by 0x6ECC26: vsg::Inherit<vsg::Group, vsg::StateGroup>::accept(vsg::RecordTraversal&) const (Inherit.h:72)
==1770777==    by 0x4DB096: void vsg::Group::t_traverse<vsg::Group const, vsg::RecordTraversal>(vsg::Group const&, vsg::RecordTraversal&) (Group.h:51)
==1770777==    by 0x4D9AD2: vsg::Group::traverse(vsg::RecordTraversal&) const (Group.h:56)
==1770777==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==1770777==
==1770777==
==1770777== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==1770777==  Access not within mapped region at address 0x0
==1770777==    at 0x70AB52: vsg::ref_ptr<vsg::GraphicsPipeline::Implementation>::operator->() const (ref_ptr.h:171)
==1770777==    by 0x70A758: vsg::GraphicsPipeline::vk(unsigned int) const (GraphicsPipeline.h:63)
==1770777==    by 0x70A5D6: vsg::BindGraphicsPipeline::record(vsg::CommandBuffer&) const (GraphicsPipeline.cpp:278)
==1770777==    by 0x7B3218: vsg::StateStack<vsg::StateCommand>::record(vsg::CommandBuffer&) (State.h:68)
==1770777==    by 0x7B2671: vsg::State::record() (State.h:292)
==1770777==    by 0x7B0484: vsg::RecordTraversal::apply(vsg::VertexIndexDraw const&) (RecordTraversal.cpp:341)
==1770777==    by 0x85F356: vsg::Inherit<vsg::Command, vsg::VertexIndexDraw>::accept(vsg::RecordTraversal&) const (Inherit.h:72)
==1770777==    by 0x6EA83D: vsg::StateGroup::traverse(vsg::RecordTraversal&) const (StateGroup.h:77)
==1770777==    by 0x7B0A65: vsg::RecordTraversal::apply(vsg::StateGroup const&) (RecordTraversal.cpp:459)
==1770777==    by 0x6ECC26: vsg::Inherit<vsg::Group, vsg::StateGroup>::accept(vsg::RecordTraversal&) const (Inherit.h:72)
==1770777==    by 0x4DB096: void vsg::Group::t_traverse<vsg::Group const, vsg::RecordTraversal>(vsg::Group const&, vsg::RecordTraversal&) (Group.h:51)
==1770777==    by 0x4D9AD2: vsg::Group::traverse(vsg::RecordTraversal&) const (Group.h:56)
```

I'm getting a nullptr access (`Address 0x0 is not...`). So I'm wondering whether I'm doing something wrong, or whether this is a bug in VSG? Should I create a minimum reproducible example?

I'm using the latest VSG Head from the git repo, commit:7ada1685.

**Comment by robertosfield** [comment]
When you create new subgraphs and add them to the scene graph once a viewer is rendering you need to run a compile traversal over the new subgraphs being adding them to the main scene graph.  The [vsgdynamicload](https://github.com/vsg-dev/vsgExamples/blob/master/examples/threading/vsgdynamicload/vsgdynamicload.cpp) example illustrates how one uses the vsg::CompileManager to compile new subgraphs, once they have been compiled.

For performance reasons the VSG doesn't have checks around low level Vulkan calls, it assumes that if an object that has Vulkan rendering associated with it then it's been precompiled before being allowed into the standard record traversal.  As users can have hundreds of thousands of Vulkan related objects in their scenes even extra if statements can cause a noticeable performance overhead.  This optimization is why you are getting a crash - the VSG is assuming the scene graph has been created and compiled correctly, but the compile is missing so the vulkan objects it's assuming should be be there aren't.

**Comment by dov** [comment]
Just wanted to say thanks. Recompiling the graph, indeed solved the problem.
